fog/view/connection/src/fog_view_router_client.rs

```diff
@@ -2,68 +2,230 @@
+
+        Ok(verification_report)
+    }
+
+    fn deattest(&mut self) {
+        if self.is_attested() {
+            log::trace!(
+                self.logger,
+                "Tearing down existing attested connection and clearing cookies."
+            );
+            self.attest_cipher = None;
+        }
+    }
+
+    /// Makes streaming requests to the fog view router service.
+    pub async fn query(
+        &mut self,
+        start_from_user_event_id: i64,
+        start_from_block_index: u64,
+        search_keys: Vec<Vec<u8>>,
+    ) -> Result<QueryResponse, Error> {
+        log::info!(self.logger, "Query was called");
+        if !self.is_attested() {
+            let verification_report = self.attest().await;
+            verification_report?;
+        }
+
+        let plaintext_request = QueryRequest {
+            get_txos: search_keys,
         };
 
-        let receive = async move {
-            let mut counter = 0;
-            while (receiver.try_next().await?).is_some() {
-                counter += 1;
-                log::info!(self.logger, "Got message {} ", counter);
-            }
-            Ok(())
+        let req_aad = QueryRequestAAD {
+            start_from_user_event_id,
+            start_from_block_index,
         };
-        let (sr, rr) = futures::join!(send, receive);
-        sr.and(rr)
+
+        let aad = mc_util_serial::encode(&req_aad);
+
+        let msg = {
+            let attest_cipher = self
+                .attest_cipher
+                .as_mut()
+                .expect("no enclave_connection even though attest succeeded");
+
+            let mut msg = Message::new();
+            msg.set_channel_id(Vec::from(attest_cipher.binding()));
+            msg.set_aad(aad.clone());
+
+            let plaintext_bytes = mc_util_serial::encode(&plaintext_request);
+
+            let request_ciphertext = attest_cipher.encrypt(&aad, &plaintext_bytes)?;
+            msg.set_data(request_ciphertext);
+            msg
+        };
+        let mut request = FogViewRouterRequest::new();
+        request.set_query(msg);
+
+        self.request_sender
+            .send((request.clone(), grpcio::WriteFlags::default()))
+            .await?;
+
+        let message = self
+            .response_receiver
+            .try_next()
+            .await?
+            .unwrap()
+            .take_query();
+
+        {
+            let attest_cipher = self
+                .attest_cipher
+                .as_mut()
+                .expect("no enclave_connection even though attest succeeded");
+
+            let plaintext_bytes = attest_cipher.decrypt(message.get_aad(), message.get_data())?;
+            let plaintext_response: QueryResponse = mc_util_serial::decode(&plaintext_bytes)?;
+            Ok(plaintext_response)
+        }
+    }
+}
+
+/// Errors related to the Fog View Router Client.
+pub enum Error {
+    /// Decode errors.
+    Decode(DecodeError),
+
+    /// Attestation errors.
+    Attestation(),
```

Comment:
:question: Is there an advantage to `Attestation()` over `Attestation`?

Reply:
No, changed.